revocation/internal/crl/crl.go

```diff
@@ -0,0 +1,306 @@
+			}
+		}
+	}
+
+	if lastErr != nil {
+		return &result.CertRevocationResult{
+			Result: result.ResultUnknown,
+			CRLResults: []*result.CRLResult{
+				{
+					Result: result.ResultUnknown,
+					URI:    crlURL,
+					Error:  lastErr,
+				}},
+		}
+	}
+
+	return &result.CertRevocationResult{
+		Result:     result.ResultOK,
+		CRLResults: crlResults,
+	}
+}
+
+// Supported checks if the certificate supports CRL.
+func Supported(cert *x509.Certificate) bool {
+	return cert != nil && len(cert.CRLDistributionPoints) > 0
+}
+
+func validate(crl *x509.RevocationList, issuer *x509.Certificate) error {
+	// check signature
+	if err := crl.CheckSignatureFrom(issuer); err != nil {
+		return fmt.Errorf("failed to verify CRL signature: %w", err)
+	}
+
+	// check validity
+	now := time.Now()
+	if !crl.NextUpdate.IsZero() && now.After(crl.NextUpdate) {
+		return fmt.Errorf("expired CRL. Current time %v is after CRL NextUpdate %v", now, crl.NextUpdate)
+	}
+
+	for _, ext := range crl.Extensions {
+		switch {
+		case ext.Id.Equal(oidFreshestCRL):
+			return ErrDeltaCRLNotSupported
+		case ext.Id.Equal(oidIssuingDistributionPoint):
+			// IssuingDistributionPoint is a critical extension that identifies
+			// the scope of the CRL. Since we will check all the CRL
+			// distribution points, it is not necessary to check this extension.
+		default:
+			if ext.Critical {
+				// unsupported critical extensions is not allowed. (See RFC 5280, Section 5.2)
+				return fmt.Errorf("unsupported critical extension found in CRL: %v", ext.Id)
+			}
+		}
+	}
+
+	return nil
+}
+
+// checkRevocation checks if the certificate is revoked or not
+func checkRevocation(cert *x509.Certificate, baseCRL *x509.RevocationList, signingTime time.Time, crlURL string) (*result.CRLResult, error) {
+	if cert == nil {
+		return nil, errors.New("certificate cannot be nil")
+	}
+
+	if baseCRL == nil {
+		return nil, errors.New("baseCRL cannot be nil")
+	}
+
+	// latestTempRevokedEntry contains the most recent revocation entry with
+	// reasons such as CertificateHold or RemoveFromCRL.
+	//
+	// If the certificate is revoked with CertificateHold, it is temporarily
+	// revoked. If the certificate is shown in the CRL with RemoveFromCRL,
+	// it is unrevoked.
+	var latestTempRevokedEntry *x509.RevocationListEntry
+	for i, revocationEntry := range baseCRL.RevokedCertificateEntries {
+		if revocationEntry.SerialNumber.Cmp(cert.SerialNumber) == 0 {
+			extensions, err := parseEntryExtensions(revocationEntry)
+			if err != nil {
+				return nil, err
+			}
+
+			// validate signingTime and invalidityDate
+			if !signingTime.IsZero() && !extensions.invalidityDate.IsZero() &&
+				signingTime.Before(extensions.invalidityDate) {
+				// signing time is before the invalidity date which means the
+				// certificate is not revoked at the time of signing.
+				continue
+			}
+
+			switch revocationEntry.ReasonCode {
+			case int(result.CRLReasonCodeCertificateHold), int(result.CRLReasonCodeRemoveFromCRL):
```

Comment:
Shouldn't `removeFromCRL` only appear in delta CRL?

Reply:
Good quesiton. According to the RFC 5280: https://www.rfc-editor.org/rfc/rfc5280#section-5.3.1
> The removeFromCRL (8) reasonCode value may only appear in delta CRLs
   and indicates that a certificate is to be removed from a CRL because
   either the certificate expired or was removed from hold.

However, I also reviewed [CA/B forum](https://cabforum.org/uploads/Baseline-Requirements-for-the-Issuance-and-Management-of-Code-Signing.v3.9.pdf)
> 7.2 CRL profile
The serial number of a revoked Certificate MUST remain on the CRL for at least 10 years after the
expiration of the Certificate.

The delta CRL will finally merge with the base CRL when the `NextUpdate` for the base CRL is reached. So, I think the entries for `CertificateHold` and `RemoveFromCRL` should be merged into the base CRL. Also, if the CA doesn't use a Delta CRL, should they update `RemoveFromCRL` directly in the base CRL?

Reply:
> I think the entries for CertificateHold and RemoveFromCRL should be merged into the base CRL. Also, if the CA doesn't use a Delta CRL, should they update RemoveFromCRL directly in the base CRL?

Can we please confirm this?

Reply:
```go
package main

import (
	"context"
	"crypto/x509"
	"fmt"
	"io"
	"net/http"
	"net/url"
)

const maxCRLSize = 32 * 1024 * 1024 // 16 MiB

// download CRL from URL
func download(ctx context.Context, crlURL string, client *http.Client) (*x509.RevocationList, error) {
	// validate URL
	parsedURL, err := url.Parse(crlURL)
	if err != nil {
		return nil, fmt.Errorf("invalid CRL URL: %w", err)
	}
	if parsedURL.Scheme != "http" {
		return nil, fmt.Errorf("unsupported CRL endpoint: %s. Only urls with HTTP scheme is supported", crlURL)
	}

	// download CRL
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, crlURL, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create CRL request %q: %w", crlURL, err)
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed for %q: %w", crlURL, err)
	}
	defer resp.Body.Close()

	// check response
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("%s %q: failed to download with status code: %d", resp.Request.Method, resp.Request.URL, resp.StatusCode)
	}

	// read with size limit
	limitedReader := io.LimitReader(resp.Body, maxCRLSize)
	data, err := io.ReadAll(limitedReader)
	if err != nil {
		return nil, fmt.Errorf("failed to read CRL response from %q: %w", resp.Request.URL, err)
	}
	if len(data) == maxCRLSize {
		return nil, fmt.Errorf("%s %q: CRL size reached the %d MiB size limit", resp.Request.Method, resp.Request.URL, maxCRLSize/1024/1024)
	}

	return x509.ParseRevocationList(data)
}

func checkTempRevokedCertificate(crl *x509.RevocationList) {
	for _, entry := range crl.RevokedCertificateEntries {
		if entry.ReasonCode == 6 {
			fmt.Printf("Certificate %s was CertificateHold\n", entry.SerialNumber)
		}

		if entry.ReasonCode == 8 {
			fmt.Printf("Certificate %s was removeFromCRL\n", entry.SerialNumber)
		}
	}
}

func main() {
	testdata := []string{
		"http://crl3.digicert.com/CloudflareIncECCCA-3.crl",
		"http://crl4.digicert.com/CloudflareIncECCCA-3.crl",
		"http://crl.r2m01.amazontrust.com/r2m01.crl",
		"http://crl.r2m02.amazontrust.com/r2m02.crl",
		"http://crl.r2m03.amazontrust.com/r2m03.crl",
		"http://c.pki.goog/wr2/oBFYYahzgVI.crl",
		"http://c.pki.goog/wr2/9UVbN0w5E6Y.crl",
		"http://c.pki.goog/wr2/75r4ZyA3vA0.crl",
		"http://c.pki.goog/wr2/GSyT1N4PBrg.crl",
		"http://c.pki.goog/we1/ygWPENklvpM.crl",
		"http://c.pki.goog/we1/H9bdJBu1Tvg.crl",
		"http://c.pki.goog/we1/T58q3x0jyXI.crl",
		"http://c.pki.goog/we1/FARePxy0M2M.crl",
		"http://c.pki.goog/we1/OhMZjmT1BfY.crl",
		"http://crl3.digicert.com/sha2-ha-server-g6.crl",
		"http://crl4.digicert.com/sha2-ha-server-g6.crl",
		"http://crl3.digicert.com/DigiCertGlobalG2TLSRSASHA2562020CA1-1.crl",
		"http://crl4.digicert.com/DigiCertGlobalG2TLSRSASHA2562020CA1-1.crl",
		"http://crl3.digicert.com/DigiCertTLSHybridECCSHA3842020CA1-1.crl",
		"http://crl4.digicert.com/DigiCertTLSHybridECCSHA3842020CA1-1.crl",
		"http://crl3.digicert.com/DigiCertTLSRSASHA2562020CA1-4.crl",
		"http://crl4.digicert.com/DigiCertTLSRSASHA2562020CA1-4.crl",
		"http://cdp.rapidssl.com/RapidSSLECCCA2018.crl",
		"http://crl3.digicert.com/DigiCertGlobalCAG2.crl",
		"http://crl4.digicert.com/DigiCertGlobalCAG2.crl",
		"http://crl3.digicert.com/DigicertSHA2SecureServerCA-1.crl",
		"http://crl4.digicert.com/DigicertSHA2SecureServerCA-1.crl",
		"http://crl3.digicert.com/DigiCertSecureSiteECCCA-1.crl",
		"http://crl4.digicert.com/DigiCertSecureSiteECCCA-1.crl",
		"http://cdp.geotrust.com/GeoTrustTLSRSACAG1.crl",
		"http://cdp.geotrust.com/GeoTrustRSACA2018.crl",
		"http://crl3.digicert.com/sha2-ev-server-g3.crl",
		"http://crl4.digicert.com/sha2-ev-server-g3.crl",
		"http://cdp.thawte.com/ThawteRSACA2018.crl",
		"http://crl3.digicert.com/DigiCertGlobalG3TLSECCSHA3842020CA1-2.crl",
		"http://crl4.digicert.com/DigiCertGlobalG3TLSECCSHA3842020CA1-2.crl",
		"http://crl3.digicert.com/DigiCertEVRSACAG2.crl",
		"http://crl4.digicert.com/DigiCertEVRSACAG2.crl",
		"http://crl.sectigo.com/SectigoRSAOrganizationValidationSecureServerCA.crl",
		"http://crl.globalsign.com/ca/gsatlasr3dvtlsca2024q1.crl",
		"http://crl.globalsign.com/gsrsaovsslca2018.crl",
		"http://crl.globalsign.com/gsgccr3dvtlsca2020.crl",
		"http://crl.globalsign.com/ca/gsatlasr3dvtlsca2023q4.crl",
		"http://crl.globalsign.com/ca/gsatlasr3dvtlsca2024q2.crl",
		"http://crl.globalsign.com/ca/gsatlasr3dvtlsca2024q3.crl",
		"http://crl.globalsign.com/gseccovsslca2018.crl",
		"http://crl.globalsign.com/alphasslcasha256g4.crl",
		"http://crl.godaddy.com/gdig2s1-8229.crl",
		"http://crl.godaddy.com/gdig2s1-14080.crl",
		"http://www.microsoft.com/pkiops/crl/Microsoft%20Azure%20RSA%20TLS%20Issuing%20CA%2007.crl",
		"http://www.microsoft.com/pkiops/crl/Microsoft%20Azure%20RSA%20TLS%20Issuing%20CA%2008.crl",
		"http://www.microsoft.com/pkiops/crl/Microsoft%20Azure%20ECC%20TLS%20Issuing%20CA%2004.crl",
		"http://crl.apple.com/apevsrsa2g1.crl",
		"http://crl.entrust.net/level1m.crl",
		"http://crl.entrust.net/level1k.crl",
		"http://crl.comodoca.com/COMODOECCOrganizationValidationSecureServerCA.crl",
	}

	for _, url := range testdata {
		crl, err := download(context.Background(), url, http.DefaultClient)
		if err != nil {
			fmt.Println(err)
			return
		}

		fmt.Println("processing CRL:", url)
		checkTempRevokedCertificate(crl)
	}
}
```
I wrote a program to check some CRLs and didn't find `CertificateHold` and `RemoveFromCRL` in the `baseCRL`. Although there is no document forbidding the addition of these two reasons in the base CRL, I cannot find an example to prove they may exist in the base CRL. Let's remove the check for these two reasons and add them back when we implement the delta CRL.

Reply:
Removed the check for `CertificateHold` and `RemoveFromCRL`.